src/VisualStudio/Core/Def/Implementation/FindReferences/VisualStudioDefinitionsAndReferencesFactory.cs

```diff
@@ -0,0 +1,176 @@
+using System.Composition;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.FindReferences;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Navigation;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.TextManager.Interop;
+using OLEServiceProvider = Microsoft.VisualStudio.OLE.Interop.IServiceProvider;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.FindReferences
+{
+    [ExportWorkspaceService(typeof(IDefinitionsAndReferencesFactory), ServiceLayer.Desktop), Shared]
+    internal class VisualStudioDefinitionsAndReferencesFactory
+        : DefaultDefinitionsAndReferencesFactory
+    {
+        private readonly IServiceProvider _serviceProvider;
+
+        [ImportingConstructor]
+        public VisualStudioDefinitionsAndReferencesFactory(SVsServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider;
+        }
+
+        protected override DefinitionItem GetThirdPartyDefinitionItem(
+            Solution solution, ISymbol definition)
+        {
+            var symbolNavigationService = solution.Workspace.Services.GetService<ISymbolNavigationService>();
+            string filePath;
+            int lineNumber, charOffset;
+            if (!symbolNavigationService.WouldNavigateToSymbol(definition, solution, out filePath, out lineNumber, out charOffset))
+            {
+                return null;
+            }
+
+            var displayParts = GetDisplayParts(filePath, lineNumber, charOffset);
+            var definitionLocation = new ExternalDefinitionLocation(
+                _serviceProvider, filePath, lineNumber, charOffset);
+
+            return new DefinitionItem(
+                GlyphTags.GetTags(definition.GetGlyph()),
+                displayParts,
+                ImmutableArray.Create<DefinitionLocation>(definitionLocation),
+                displayIfNoReferences: true);
+        }
+
+        private ImmutableArray<TaggedText> GetDisplayParts(
+            string filePath, int lineNumber, int charOffset)
+        {
+            var builder = ImmutableArray.CreateBuilder<TaggedText>();
+
+            var sourceLine = GetSourceLine(filePath, lineNumber).Trim(' ', '\t');
+
+            // Put the line in 1-based for the presentation of this item.
+            var formatted = $"{filePath} - ({lineNumber + 1}, {charOffset + 1}) : {sourceLine}";
+
+            return ImmutableArray.Create(new TaggedText(TextTags.Text, formatted));
+        }
+
+        private string GetSourceLine(string filePath, int lineNumber)
+        {
+            using (var invisibleEditor = new InvisibleEditor(
+                _serviceProvider, filePath, needsSave: false, needsUndoDisabled: false))
```

Comment:
1.  this is for external items.  I have no idea if we can even read these off disk as they may not exist.
2. This completely preserves the existing code.  I did not write this and i do *not* want to break something in an area i'm not sure anyone on our team is that familiar with :)